### PR TITLE
Do not raise errors when attempting to fetch malformed ITU-R refs

### LINKED ISSUE #56: Unable to fetch "ITU-R Recommendation 592"
Attempt to fetch `ITU-R Recommendation 592` with Relaton results with an error. It is a regression — it was working without error on 18th March. (I'm not 100% sure if the return value was correct, but there was no error.)

Exact error:

```
`split': bad URI(is not URI?): "https://raw.githubusercontent.com/relaton/relaton-data-itu-r/master/data/RECOMMENDATION 592.yaml" (URI::InvalidURIError)
```

Stack trace:

```
	14: from (redacted)/gems/relaton-1.7.7/lib/relaton/db.rb:73:in `fetch'
	13: from (redacted)/gems/relaton-1.7.7/lib/relaton/db.rb:384:in `check_bibliocache'
	12: from (redacted)/gems/relaton-1.7.7/lib/relaton/db.rb:404:in `new_bib_entry'
	11: from (redacted)/gems/relaton-1.7.7/lib/relaton/db.rb:419:in `net_retry'
	10: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/processor.rb:17:in `get'
	 9: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/itu_bibliography.rb:49:in `get'
	 8: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/itu_bibliography.rb:137:in `itubib_get1'
	 7: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/itu_bibliography.rb:91:in `search_filter'
	 6: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/itu_bibliography.rb:28:in `search'
	 5: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/itu_bibliography.rb:28:in `new'
	 4: from (redacted)/gems/relaton-itu-1.7.7/lib/relaton_itu/hit_collection.rb:30:in `initialize'
	 3: from (redacted)/lib/ruby/2.7.0/uri/common.rb:737:in `URI'
	 2: from (redacted)/lib/ruby/2.7.0/uri/common.rb:234:in `parse'
	 1: from (redacted)/lib/ruby/2.7.0/uri/rfc3986_parser.rb:73:in `parse'
```

Recorded Pry session:

```
[2] pry(main)> require "relaton"
=> true
[3] pry(main)> rec = "ITU-R Recommendation 592"
=> "ITU-R Recommendation 592"
[4] pry(main)> Relaton::Db.new("db1", nil).fetch(rec)
[relaton] Info: detecting backends:
[relaton] processor "relaton_gb" registered
[relaton] processor "relaton_iec" registered
[relaton] processor "relaton_ietf" registered
[relaton] processor "relaton_iso" registered
[relaton] processor "relaton_itu" registered
[relaton] processor "relaton_nist" registered
[relaton] processor "relaton_ogc" registered
[relaton] processor "relaton_calconnect" registered
[relaton] processor "relaton_omg" registered
[relaton] processor "relaton_un" registered
[relaton] processor "relaton_w3c" registered
[relaton] processor "relaton_ieee" registered
[relaton] processor "relaton_iho" registered
[relaton] processor "relaton_bipm" registered
[relaton] processor "relaton_ecma" registered
[relaton] processor "relaton_cie" registered
[relaton-itu] ("ITU-R Recommendation 592") fetching...
URI::InvalidURIError: bad URI(is not URI?): "https://raw.githubusercontent.com/relaton/relaton-data-itu-r/master/data/RECOMMENDATION 592.yaml"
from /Users/skale/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/rfc3986_parser.rb:67:in `split'
```

</details>

### PR DESCRIPTION
**Pull request by skalee**
This pull request brings an improved regular expression for matching ITU-R refs, which disallows malformed refs. In the result, `URI::InvalidURIError` is not raised any more. Fixes #56.

### CODE REVIEW

File: lib/relaton_itu/hit_collection.rb [hunk 1]
```diff
@@ -24,8 +24,8 @@ def initialize(ref, year = nil) # rubocop:disable Metrics/AbcSize,Metrics/Method
         resp = Net::HTTP.post(uri, data.to_json,
                               "Content-Type" => "application/json")
         @array = hits JSON.parse(resp.body)
-      elsif ref.match? /^ITU-R/
-        rf = ref.sub(/^ITU-R\s/, "").upcase
+      elsif ref.match /^ITU-R\s([-_.\w]+)$/
```
[the remark this answers is not in the document]
Comment (skalee) [reply lib/relaton_itu/hit_collection.rb:27]: I'm going to prepare a pull request with some refactoring of this class anyway.

**Comment by skalee** [comment]
@andrew2net, it looks like we did it exactly in parallel.

**Comment by andrew2net** [comment]
@skalee not exactly. I've added catching an `URI::InvalidURIError` and raising a `RelatonBib::RequestError`.
The difference between raising `RelatonBib::RequestError` and returning `nil` is in the `relaton`'s behavior.
In case the `relaton` gets `nil` from flavor gem it considers that request was completed correctly but a document wasn't found. In this case, `relaton` creates a "not found record" in a cache for reference. The `relaton` won't make requests for the document reference until the cache is valid.
If `RelatonBib::RequestError` is raised it means that something goes wrong and a record in the cache won't be made.

**Comment by skalee** [comment]
@andrew2net  I don't understand why requesting ref `ITU-R Recommendation 592` raises exception whilst requesting `ITU-R Recommendation592` returns `nil`. It feels inconsistent. IMO in both these cases the result should be either `nil` (document not found) or some exception (clearly invalid ref).

Furthermore, caching `URI::InvalidURIError` means that actually you're caching Relaton internal error instead of fixing the bug — either by improving the regular expression so that it disallows spaces or by escaping that ref (`CGI.escape`) before injecting it to URI).

**Comment by andrew2net** [comment]
> I don't understand why requesting ref `ITU-R Recommendation 592` raises exception whilst requesting `ITU-R Recommendation592` returns `nil`. It feels inconsistent. IMO in both these cases the result should be either `nil`

@skalee I agree, it should be `nil` for both cases. I just said that I did not exactly the same thing and tried to explain how the communications between gems work. So please finish your work and merge the PR.

> Furthermore, caching `URI::InvalidURIError` means that actually you're caching Relaton internal error instead of fixing the bug — either by improving the regular expression so that it disallows spaces or by escaping that ref (`CGI.escape`) before injecting it to URI.

We need to continue running `relaton` and `relaton-cli` gems even some kind of exceptions happen in the flavor gems like `relaton-itu`. So we use `RelatonBib::ReauestError` to tell `relaton` and `relaton-cli` gems that something went wrong but they should continue fetching other references.

**Comment by skalee** [comment]
> So please finish your work and merge the PR.

It is complete in terms of the bugfix. Regarding tests, they failed because tests on master were failing too. I'll rebase in a moment and they should be all green.

Unless you prefer `CGI.escape` instead of making the regexp more strict. Both solutions are equally easy.

**Comment by andrew2net** [comment]
Thanks, @skalee. Could you fix the Hound's offence?

**Comment by skalee** [comment]
> Could you fix the Hound's offence?

I wanted to do clean-ups like that in a separate pull request, because it required some refactoring, but I've just added these commits here.

**Pull request merged**